repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -0,0 +1,619 @@
+import org.openrepose.core.services.config.ConfigurationService
+import org.openrepose.core.services.datastore.{Datastore, DatastoreService}
+import org.openrepose.core.services.serviceclient.akka.{AkkaServiceClient, AkkaServiceClientException}
+import org.openrepose.core.systemmodel.SystemModel
+import org.openrepose.filters.keystonev2.KeystoneRequestHandler._
+import org.openrepose.filters.keystonev2.config._
+
+import scala.collection.JavaConverters._
+import scala.language.implicitConversions
+import scala.util.{Failure, Random, Success, Try}
+
+@Named
+class KeystoneV2Filter @Inject()(configurationService: ConfigurationService,
+                                 akkaServiceClient: AkkaServiceClient,
+                                 datastoreService: DatastoreService)
+  extends Filter
+  with HttpDelegationManager
+  with LazyLogging {
+
+  import KeystoneV2Filter._
+
+  private var configurationFile: String = DEFAULT_CONFIG
+  private var sendTraceHeader = true
+
+  private val datastore: Datastore = datastoreService.getDefaultDatastore //Which happens to be the local datastore
+
+  var keystoneV2Config: KeystoneV2Config = _
+
+  override def init(filterConfig: FilterConfig): Unit = {
+    configurationFile = new FilterConfigHelper(filterConfig).getFilterConfig(DEFAULT_CONFIG)
+    logger.info(s"Initializing Keystone V2 Filter using config $configurationFile")
+    val xsdURL: URL = getClass.getResource("/META-INF/schema/config/keystone-v2.xsd")
+    configurationService.subscribeTo(
+      SYSTEM_MODEL_CONFIG,
+      getClass.getResource("/META-INF/schema/system-model/system-model.xsd"),
+      SystemModelConfigListener,
+      classOf[SystemModel]
+    )
+    configurationService.subscribeTo(
+      filterConfig.getFilterName,
+      configurationFile,
+      xsdURL,
+      KeystoneV2ConfigListener,
+      classOf[KeystoneV2Config]
+    )
+  }
+
+  override def destroy(): Unit = {
+    configurationService.unsubscribeFrom(configurationFile, KeystoneV2ConfigListener)
+    configurationService.unsubscribeFrom(SYSTEM_MODEL_CONFIG, SystemModelConfigListener)
+  }
+
+  override def doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, chain: FilterChain): Unit = {
+    /**
+     * STATIC REFERENCE TO CONFIG
+     */
+    val config = keystoneV2Config
+
+    /**
+     * DECLARE COMMON VALUES
+     */
+    lazy val request = MutableHttpServletRequest.wrap(servletRequest.asInstanceOf[HttpServletRequest])
+    // Not using the mutable wrapper because it doesn't work properly at the moment, and
+    // we don't need to modify the response from further down the chain
+    lazy val response = servletResponse.asInstanceOf[HttpServletResponse]
+    lazy val traceId = Option(request.getHeader(CommonHttpHeader.TRACE_GUID.toString)).filter(_ => sendTraceHeader)
+    lazy val requestHandler = new KeystoneRequestHandler(getIdentityServiceUri(config.getIdentityService.getUri), akkaServiceClient, traceId)
```

Comment:
I kind of don't like computing this twice for every request when it really only changes when the configuration is loaded. Can we move this logic into the configurationUpdated callback?